File: ta_lib/strategies/signal/src/signalline/qstick.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use trend::qstick;

pub struct QstickSignalLineSignal {
    smooth_type: Smooth,
    period: usize,
    signal_period: usize,
}

impl QstickSignalLineSignal {
    pub fn new(smooth_type: Smooth, period: f32, signal_period: f32) -> Self {
        Self {
            smooth_type,
            period: period as usize,
            signal_period: signal_period as usize,
        }
    }
}

impl Signal for QstickSignalLineSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.signal_period)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let qstick = qstick(data.open(), data.close(), self.smooth_type, self.period);
        let signal_line = qstick.smooth(self.smooth_type, self.signal_period);

        (
            qstick.cross_over(&signal_line),
            qstick.cross_under(&signal_line),
        )
    }
}

File: strategy/generator/signal/signalline/qstick.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class QstickSignalLineSignal(Signal):
    type: SignalType = SignalType.QstickSignalLine
    smooth_type: Parameter = StaticParameter(Smooth.EMA)
    period: Parameter = RandomParameter(10.0, 15.0, 1.0)
    signal_period: Parameter = RandomParameter(4.0, 8.0, 1.0)
